src/NHibernate/Util/ExpressionsHelper.cs

```diff
@@ -360,25 +360,7 @@ internal static IType GetType(VisitorParameters parameters, Expression expressio
 						break;
 					case IAbstractComponentType componentType:
 						currentComponentType = componentType;
-						if (currentEntityPersister == null)
-						{
-							// When persister is not available (q.OneToManyCompositeElement[0].Prop), try to get the type from the component
-							currentType = TryGetComponentPropertyType(componentType, member.Path);
-						}
-						else
-						{
-							// Concatenate the component property path in order to be able to use EntityMetamodel.GetPropertyType to retrieve the type.
```

Comment:
EntityMetamodel also can't handle ICustomUserType components. So why bother to use it at all if we can retrieve all info from `componentType`

Reply:
> Concatenate the component property path in order to be able to use EntityMetamodel.GetPropertyType

From original comments it looks like it's some internal calculations specifically for EntityMetamodel call. But this calculated path is returned in out param `memberPath`.

@maca88 Are you sure this concatenation is not required for the branch "When persister is not available".
You have it covered in the following test:
https://github.com/nhibernate/nhibernate-core/blob/a97627d8f1b06388aa68d4b0c5ddf386a0a1781b/src/NHibernate.Test/Linq/TryGetMappedTests.cs#L461-L466

Where you expect `memberPath` for `Select(o => o.Components[0].Subcomponent.Name)` to be `Name`. But why? Shouldn't it be `Subcomonent.Name` instead?

Also should such query `session.Query<Baz>().Select(o => o.Components[0].Subcomponent.Name).ToList()` work at all? Currently it fails to parse generated hql...

Reply:
> But why? Shouldn't it be Subcomonent.Name instead?

I made a quick look and it seems not important. Both cases are handled.